### PR TITLE
Exit code change BZ #2052697

### PR DESCRIPTION
**Pull request by jwhonce**
* On SIGTERM, podman service will return exit code 0. SIGKILL will return exit code 1. This better matches expected behavior from `systemctl stop podman.service`
* Add e2e test to cover case above plus ensuring containers are still running after podman.service is stopped via `systemctl stop podman.service`

See https://bugzilla.redhat.com/show_bug.cgi?id=2052697

Signed-off-by: Jhon Honce <jhonce@redhat.com>

### CODE REVIEW

**Review by edsantiago — CHANGES_REQUESTED** [review]
Note that tests are hanging. I haven't tried reproducing the hang, will leave that up to you, but I have some suggestions inline.

File: test/apiv2/80-systemd.at [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+ACTIVATE=/usr/bin/systemd-socket-activate
+if ! [[ -x $ACTIVATE ]]; then
+    echo "skip: $(basename $0) tests require $ACTIVATE" >&2
+    exit
+fi
+
+# systemd-socket-activate does not support DNS lookups
+OLDHOST=$HOST
+HOST=127.0.0.1
```
[the remark this answers is not in the document]
Comment (jwhonce) [reply test/apiv2/80-systemd.at:18]: The function t() uses `HOST` and in a previous iteration 80-systemd.at depended on it. I can remove that now.

File: test/apiv2/80-systemd.at [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+wait -n $ACTIVATE_PID || rc=$?
+is $rc 1 "API service return code"
+
+running=$($PODMAN_BIN --root $WORKDIR/server_root inspect --format '{{.State.Running}}' $cid)
+is "$running" "true" "Container running state"
+
+HOST=$OLDHOST
+kill -KILL $ACTIVATE_PID 2>/dev/null ||:
+podman rm -af ||:
+podman rmi -af ||:
```
[the remark this answers is not in the document]
Comment (jwhonce) [reply test/apiv2/80-systemd.at:60]: ack, C&P from test-apiv2

File: test/apiv2/80-systemd.at [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+#
+
+## Verify containers created under API service are properly configured
+## See https://bugzilla.redhat.com/show_bug.cgi?id=2052697
+
+
+ACTIVATE=/usr/bin/systemd-socket-activate
+if ! [[ -x $ACTIVATE ]]; then
+    echo "skip: $(basename $0) tests require $ACTIVATE" >&2
+    exit
```
[the remark this answers is not in the document]
Comment (jwhonce) [reply test/apiv2/80-systemd.at:13]: Will change

File: test/apiv2/80-systemd.at [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+fi
+
+# systemd-socket-activate does not support DNS lookups
+OLDHOST=$HOST
+HOST=127.0.0.1
+
+ACTIVATE_PORT=$(random_port)
+if [[ $PORT == $ACTIVATE_PORT ]]; then
+    die "Duplicate port($PORT) returned for socket activation tests"
+fi
```
[the remark this answers is not in the document]
Comment (edsantiago) [reply test/apiv2/80-systemd.at:23]: It does hurt, in the sense that a future maintainer will read that and wonder why the check is there, or how likely that is to happen, and if it is, why isn't it fixed in `random_port()` itself, and maybe waste time looking at the function and trying to fix it (but it's already fixed). If you choose to keep it, could you add a comment such as "this is impossible, but I'm paranoid"?

File: test/apiv2/80-systemd.at [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+    .Id~[0-9a-f]\\{64\\}
+cid=$(jq -r '.Id' <<<"$output")
+t POST "$ACTIVATE_URL/containers/$cid/start" 204
+
+t GET "$ACTIVATE_URL/containers/$cid/json" 200 \
+      .State.Running=true
+
+# Emulate 'systemd stop podman.service'
+kill -TERM $ACTIVATE_PID
+wait -n $ACTIVATE_PID || rc=$?
```
Comment (edsantiago) [inline test/apiv2/80-systemd.at:51]: `-n` isn't needed here, hence it's misleading to a future maintainer.

I wonder if this is the cause of the hang?

File: test/apiv2/test-apiv2 [hunk 6]
```diff
@@ -194,8 +194,14 @@ function jsonify() {
...
 
-        # If right-hand side already includes double quotes, do nothing
-        if [[ ! $rhs =~ \" ]]; then
+        # If right-hand side already includes double quotes or boolean value, do nothing
+        if [[ $rhs =~ \" || $rhs =~ true || $rhs =~ false ]]; then
+            true
+        elif [[ $rhs == False ]]; then
+            rhs=false
+        elif [[ $rhs == True ]]; then
+            rhs=true
```
[the remark this answers is not in the document]
Comment (edsantiago) [reply test/apiv2/test-apiv2:203]: I see this marked as resolved, but not actually resolved??

File: test/apiv2/80-systemd.at (hunk 5 above)
Comment (jwhonce) [reply test/apiv2/80-systemd.at]: I added timeout to see.

File: test/apiv2/test-apiv2 (hunk 6 above)
[the remark this answers is not in the document]
Comment (jwhonce) [reply test/apiv2/test-apiv2:203]: I updated comment to be more explicate, also added test for numeric values. Will show up in next commit.

File: test/apiv2/80-systemd.at (hunk 4 above)
[the remark this answers is not in the document]
Comment (jwhonce) [reply test/apiv2/80-systemd.at]: I removed it, we can also add it back if needed.

File: libpod/runtime.go [hunk 7]
```diff
@@ -210,6 +210,10 @@ func newRuntimeFromConfig(ctx context.Context, conf *config.Config, options ...R
 	}
 
 	if err := shutdown.Register("libpod", func(sig os.Signal) error {
+		// For `systemctl stop podman.service` support, exit code should be 0
+		if sig == syscall.SIGTERM {
```
Comment (mheon) [inline libpod/runtime.go:214]: This doesn't seem right - I'm trying a lot of longrunning commands and hitting them all with SIGTERM, and none of them exit cleanly?

Comment (jwhonce) [reply libpod/runtime.go:214]: `systemctl stop podman.service` always returned 1, after this change it now returns 0 as expected.

**Comment by mheon** [comment]
LGTM

**Pull request merged**